### PR TITLE
Handle leading delimiter char in axom::Path

### LINKED ISSUE #777: Preserve leading delimiter in axom::Path
The current implementation of axom::Path stores a vector of strings, the components of the path, along with a single delimiter character, usually `/`.  The result of the implementation is that there is no way to construct a `Path` that has a leading delimiter.  The following two `Path` objects will be identical, and no methods of the class can retrieve the fact that `path_1` was constructed using a string with a leading `/`.

` Path path_0("foo/bar/baz");`
` Path path_1("/foo/bar/baz");`

The leading `/` is especially important when the `Path` is used to represent a file system path, so we should modify the class to preserve the leading delimiter and use it in the various methods of the class.

### PR DESCRIPTION
**Pull request by nselliott**
# Summary

- This PR is a bug fix
- It does the following (modify list as needed):
  - Modifies class `axom::Path` to preserve the a leading delimiter in the path, such as `/foo/bar`.
  - Fixes #777

Reviewers, please especially take a look at the unit tests added here, to be sure that the expected results for the tests match the desired functionality of the class. The handling of paths such as the root path `/` and those with only one component like `/foo` could be considered new functionality.

### CODE REVIEW

File: src/axom/core/tests/core_Path.hpp [hunk 1]
```diff
@@ -116,15 +188,25 @@ TEST(core_Path, dirname_basename_multi_component_alt_delim)
...
+  axom::Path path("/foo/bar/baz");
+  EXPECT_EQ(path.dirName(), "/foo/bar");
+  EXPECT_EQ(path.baseName(), "baz");
+}
+
 TEST(core_Path, basic_equals)
 {
   axom::Path first("foo/bar/baz");
   axom::Path second("foo/bar/baz");
   axom::Path third("foo:bar:baz", ':');
```
Comment (kennyweiss) [inline src/axom/core/tests/core_Path.hpp:202]: Suggestion: We should probably also test the leading delimeter with a non-default delim, 
e.g. `Path {":foo:bar:baz", ':' }`

File: src/axom/core/Path.cpp [hunk 2]
```diff
@@ -45,12 +62,29 @@ Path Path::join(std::initializer_list<Path> paths, const char delim)
...
   return result;
 }
 
 Path::operator std::string() const
 {
-  return fmt::format("{0}", fmt::join(m_components, std::string(1, m_delim)));
+  std::string result;
+  if(m_leading_delim)
+  {
+    result = "/";
```
Comment (kennyweiss) [inline src/axom/core/Path.cpp:83]: Should this account for a non-default delimiter?
E.g.
```suggestion
    result = m_delim;
```

File: src/axom/core/Path.cpp [hunk 3]
```diff
@@ -45,12 +62,29 @@ Path Path::join(std::initializer_list<Path> paths, const char delim)
...
+  }
+
+  result += fmt::format("{0}", fmt::join(m_components, std::string(1, m_delim)));
```
Comment (kennyweiss) [inline src/axom/core/Path.cpp:86]: Suggestion: The following one-liner could account for leading delims and joining the path
```suggestion
  return fmt::format("{0}{1}", 
                 m_leading_delim? "" : std::string(1,m_delim), 
                 fmt::join(m_components, std::string(1, m_delim)));
```

File: src/axom/core/Path.cpp [hunk 4]
```diff
@@ -16,8 +16,15 @@ namespace axom
 {
 Path::Path(const std::string& path, const char delim) : m_delim(delim)
 {
+  size_t first_position = 0;  // position of first non-delimiter char in path
+  if(!path.empty() && path[0] == m_delim)
```
Comment (kennyweiss) [inline src/axom/core/Path.cpp:20]: Suggestion: Consider using `axom::utilities::string::startsWith()` for this test

https://github.com/LLNL/axom/blob/cd5866a630ada425e9f0f0de0c54c4708543f48c/src/axom/core/utilities/StringUtilities.hpp#L55-L64

**Review by kennyweiss — APPROVED** [review]
Thanks @nselliott !

File: src/axom/core/Path.cpp (hunk 2 above)
Comment (nselliott) [reply src/axom/core/Path.cpp:83]: Thanks for catching that, I overlooked this one.

File: src/axom/core/tests/core_Path.hpp [hunk 5]
```diff
@@ -14,6 +14,13 @@ TEST(core_Path, trivial_construct)
   EXPECT_EQ(static_cast<std::string>(path), path_str);
 }
 
+TEST(core_Path, lead_delim_construct)
+{
+  const std::string path_str = "/foo";
+  axom::Path path(path_str);
+  EXPECT_EQ(static_cast<std::string>(path), path_str);
```
[the remark this answers is not in the document]
Comment (nselliott) [reply src/axom/core/tests/core_Path.hpp:21]: The cast uses an implemented `operator std::string()`, so it shouldn't break as long as the implementation continues to be tested and maintained. This was an existing part of the API that I didn't touch. If we want an API change, perhaps that should be a separate issue?

[the remark this answers is not in the document]
Comment (agcapps) [reply src/axom/core/tests/core_Path.hpp:21]: Thanks for pointing out the `operator std::string()`; I had overlooked that.  I don't want to inflict scope creep on this PR or on you!  Perhaps a separate issue would be a good idea.

File: src/axom/core/Path.cpp (hunk 3 above)
Comment (nselliott) [reply src/axom/core/Path.cpp:86]: Thanks, I have put this in.  I needed to reverse the order of the second and third operands in the ternary operator.

**Review by rhornung67 — APPROVED** [review]

**Review by white238 — APPROVED** [review]

**Pull request merged**